# HAF-pipe ðŸ‚

HAF-pipe is a bash- and R-based pipeline to calculate haplotype-inferred allele frequencies from pool-seq data and founder SNP profiles.

## Paper



> Tilk et al. High accuracy haplotype-derived allele frequencies from ultra-low coverage pool-seq samples, bioRxiv, 2018. https://www.biorxiv.org/content/early/2018/01/11/244004  

## Dependencies



1. HARP (see https://bitbucket.org/dkessner/harp )
2. R version >= 3.2 (in your path)

## Usage



```
HAFpipe-line.sh  [ -b bamfile ] | [ -bl filename of bamfile list ] 
                    [ -r reference sequence ]
                    [ -v vcf ] | [ -t snp table ]
                    [ -s steps to run ] 
                    [ -c chromosomes to run ]
                    [ -o output directory ]
                    [ -f founder subset list (default none) ]
                    [ -e base quality encoding (default illumina) ]
                    [ -w window size (in kb) for haplotype inference (default 1000kb) ]
                    [ -t number of threads to use for parallization  (default 1) ]
                    [ -h help ]
```
## Required Parameters



Each parameter requires the full path to be specified.

 * **-b**  : a bam file generated by pooled sequencing of sampled chromosomes from an evolved population
 * **-r**  : a reference sequence in FASTA format
 * **-bl** : a text file of a list of bam files, where each line contains the full path of a bam file
 * **-v**  : a multi-sample VCF file of each individual founder in the starting population 
 * **-t**  : a SNP table that contains the the position, ref allele, and genotypes of the each individual founder. See ``` simulations/99.clean.SNP.HARP.segregating.gz``` included in this repository for examples.
 * **-s**  : Numerical value of the step(s) in the pipeline to run separated by a comma <br>
   * 1 - generates a SNP table from a multi-sample VCF file <br>
   * 2 - imputes missing sites from SNP table <br>
   * 3 - infer haplotype frequencies <br>
   * 4 - calculate haplotype-derived allele frequencies (HAFs) <br>
 * **-c**  : name of chromosome  <br>
 * **-o**  : Output directory <br>

## Optional Parameters



 * **-f**  : a text file of each founder to include in the SNP table, where each line contains the name of the founder in the original SNP table (default none)
 * **-w** : size window (in kb) for haplotype inference (default 1000) 
 * **-e** : base quality encoding, illumina (Phred+64) or sanger (Phred+33) (default illumina)
 * **-t**  : number of threads to use (default 1)

## Example


Estimates HAFs from multiple pool-seq samples  <br>
``` wrapper.sh -bl bamsList.txt -r dmelRefrence.fa -v founderSamples.vcf -s 1,2,3,4 -c 2L -o /home/project ```
